tests/unit/memefficiency.tcl

```diff
@@ -586,6 +586,100 @@ run_solo {defrag} {
             }
         }
 
+        test "Active defrag for argv retained by the main thread from IO thread: $type" {
+            r flushdb
+            r config set hz 100
+            r config set activedefrag no
+            wait_for_defrag_stop 500 100
+            r config resetstat
+            if {[lindex [r config get io-threads] 1] == 1} {
+                r config set active-defrag-threshold-lower 5
+            } else {
+                r config set active-defrag-threshold-lower 10
+            }
+            r config set active-defrag-cycle-min 65
+            r config set active-defrag-cycle-max 75
+            r config set active-defrag-ignore-bytes 1000kb
+            r config set maxmemory 0
+
+            # Create some clients so that they are distributed among different io threads.
+            set clients {}
+            for {set i 0} {$i < 32} {incr i} {
+                lappend clients [redis_client]
+            }
+
+            # Populate memory with interleaving key pattern of same size
+            set dummy "--[string repeat x 400]\nreturn "
+            set n 50000
+            for {set j 0} {$j < $n} {incr j} {
+                set val "$dummy[format "%06d" $j]"
+                set rr [lindex $clients [expr {int(rand() * 32)}]]
+                $rr set k$j $val
+                $rr set f$j $val
+            }
+
+            for {set i 0} {$i < 32} {incr i} {
+                [lindex $clients $i] close
+            }
+            
+            after 120 ;# serverCron only updates the info once in 100ms
+            if {$::verbose} {
+                puts "used [s allocator_allocated]"
+                puts "rss [s allocator_active]"
+                puts "frag [s allocator_frag_ratio]"
+                puts "frag_bytes [s allocator_frag_bytes]"
+            }
+            assert_lessthan [s allocator_frag_ratio] 1.05
+
+            # Delete all the keys to create fragmentation
+            set rd [redis_deferring_client]
+            for {set j 0} {$j < $n} {incr j} { $rd del f$j }
+            for {set j 0} {$j < $n} {incr j} { $rd read } ; # Discard del replies
+            $rd close
```

Comment:
deleting **all** the keys doesn't cause fragmentation. you need to delete interleaved keys that were created in the same arena.

i'd suggest this:
1. instead of random client, and random order, let's let each client create 10000 keys sequentially (before moving to the next client. where each client uses a different key prefix, not to collide with other clients.
2. add some code to verify that indeed the allocations that were retained are not in the main arena.
3. delete all the even keys for each of the prefixes.
4. we better choose a jemalloc bucket size that is prone to fragmentation (one with lots of registers in each slab).

Reply:
> deleting **all** the keys doesn't cause fragmentation. 

the current test didn't delete all the keys, i created keys with the prefix `k` and `f` interchangeably, and ended up removing only keys with the prefix `f`.
but the way you suggested made more sense.

> i'd suggest this:
> 
> 1. instead of random client, and random order, let's let each client create 10000 keys sequentially (before moving to the next client. where each client uses a different key prefix, not to collide with other clients.
> 2. add some code to verify that indeed the allocations that were retained are not in the main arena.
> 3. delete all the even keys for each of the prefixes.
> 4. we better choose a jemalloc bucket size that is prone to fragmentation (one with lots of registers in each slab).

good ideas.

Reply:
done with [`e50a183` (#13665)](https://github.com/redis/redis/pull/13665/commits/e50a183fc080ec4dab41a01f41288c8fc391a531)